File: DomainSpaceFrontend/src/types/common/PagedList.ts
export class PagedList<T> {
  pageNumber?: number;
  pageSize?: number;
  totalCount?: number;
  lastPage?: number;
  items?: T[];

  constructor({
    pageNumber,
    pageSize,
    totalCount,
    lastPage,
    items,
  }: PagedList<T> = {}) {
    this.pageNumber = pageNumber;
    this.pageSize = pageSize;
    this.totalCount = totalCount;
    this.lastPage = lastPage;
    this.items = items;
  }

  static fromJson<T>(
    data: any,
    itemConstructor: { fromJson: (json: any) => T }
  ): PagedList<T> {
    return new PagedList<T>({
      pageNumber: data["pageNumber"],
      pageSize: data["pageSize"],
      items: data["items"].map((item: any) => itemConstructor.fromJson(item)),
      lastPage: data["lastPage"],
      totalCount: data["totalCount"],
    });
  }
}

File: DomainSpaceBackend/DomainSpace.Common/Dto/PagedList.cs
﻿namespace DomainSpace.Common.Dto;

/// <summary>
/// Paged list
/// </summary>
/// <typeparam name="T">Type</typeparam>
public class PagedList<T>
{
    /// <summary>
    /// Page number
    /// </summary>
    public int PageNumber { get; set; }

    /// <summary>
    /// Page size
    /// </summary>
    public int PageSize { get; set; }

    /// <summary>
    /// Result
    /// </summary>
    public IEnumerable<T>? Items { get; set; }

    /// <summary>
    /// Total count
    /// </summary>
    public int TotalCount { get; set; }

    /// <summary>
    /// LastPage
    /// </summary>
    public int LastPage
    {
        get
        {
            if (TotalCount <= 0 || PageSize <= 0)
            {
                return 1;
            }

            return (int)Math.Ceiling(1.0 * TotalCount / PageSize);
        }
    }
}
